created Session class, added tests for update user endpoint

yuliiahonchar opened the pull request.

yuliiahonchar committed:
added positive tests for update user endpoint, added session methods, added executes methods into Response Handler class, renamed email constants

yuliiahonchar committed:
added testNG xml

vhonchar commented on src/main/java/org/realworld/api/Session.java:
```diff
@@ -1,20 +1,63 @@
         return this.user.getUser().getEmail();
     }
+
+    public Session createSessionWithEmail(String email, ApiService serviceClass) {
```
I would suggest to separate session creation process from session storage. 
This `Session` class looks like a simple POJO to store information about current user's session and in its current state, it looks a bit complicated because of dependency on `ApiService`. If `createSessionWithEmail` was just creating an instance of Session with provided parameters, it would be ok to have some **static** method in a POJO, there is no complex logic in that, just composing a Session from different parameters. 

But, since `createSessionWithEmail` is a complex method with some logic AND dependency on another service (`ApiService`), this can already be treated as "complex logic for a new class"

Better to have some `SessionService` or `AuthenticationService`, which would be responsible for creating an instance of Session (and doing all this sign-in/sign-up logic)

yuliiahonchar replied:
done

vhonchar commented on src/main/java/org/realworld/utils/ResponseHandler.java:
```diff
@@ -0,0 +1,74 @@
+        return getParsedErrorResponse(rawResponse);
+    }
+
+    public static <T> ResponseWrapper<T> getParsedResponse(Response<T> rawResponse) {
```
**suggestion**: to make name shorter, you can use just `parse`.  This method has only one parameter with certain type Response, so no necessity to mention it in the name, and usage of `get` everywhere makes it somewhat useless (this method doesn't get something from somewhere, right?)

yuliiahonchar replied:
done

vhonchar commented on src/main/java/org/realworld/utils/ResponseHandler.java:
```diff
@@ -0,0 +1,74 @@
+        return new ResponseWrapper<>(rawResponse.code(), rawResponse.body());
+    }
+
+    public static <T> ResponseWrapper<String> getParsedErrorResponse(Response<T> rawResponse) {
```
**suggestion**: same as above, can be just `parseError`

yuliiahonchar replied:
done

vhonchar commented on src/main/java/org/realworld/utils/ResponseHandler.java:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    public static <T> Response<T> executeSuccessfully(Call<T> request) {
+        Response<T> response;
```
**clean code**: no point in creating a variable here. It can be done inside of the try block, like:
```suggestion
try {
    Response<T> response = request.execute();
    if (!response.isSuccessful()) {
        throw new RuntimeException("Unexpected status code for this request: " + response.code());
     }
     return response
}    
```

The same in `execute` method

yuliiahonchar replied:
done

vhonchar commented on src/main/java/org/realworld/utils/ResponseHandler.java:
```diff
@@ -0,0 +1,74 @@
+
+import java.io.IOException;
+
+public class ResponseHandler {
```
**suggestion**: Handler, mostly reverse to a class which should be instantiated to handle something. More-over, usually, a pattern is to have abstract class or interface `ResponseHandler` and then have specific implementations, since both words are too generic. 

In this case: a class with bunch of public static methods, usually called `SomethingUtils`

yuliiahonchar replied:
done

vhonchar commented on src/main/resources/apiTestSuites/fullApiRegression.xml:
```diff
@@ -0,0 +1,19 @@
+<!DOCTYPE suite SYSTEM "https://testng.org/testng-1.0.dtd" >
```
shouldn't this file be under `src/test/resource/...`? Since it is related to tests

yuliiahonchar replied:
done

vhonchar commented on src/test/java/api/signup/UpdateUserPositiveTest.java:
```diff
@@ -0,0 +1,150 @@
+    private ApiService authorizedApiService;
+
+    @BeforeMethod(alwaysRun = true)
+    public void SignIn() {
```
is it common to name methods with capital letter in TestNG? Because usually, methods in Java a starting from lower case

yuliiahonchar replied:
done

vhonchar commented on src/test/java/api/signup/UpdateUserPositiveTest.java:
```diff
@@ -0,0 +1,150 @@
+        sAssert.assertEquals(parsedResponse.getStatusCode(), 200);
+        sAssert.assertEquals(parsedResponse.getResponseBody().getUser().getUsername(), updatedUsername);
+        sAssert.assertAll();
+    }
```
**suggestion**: add a case to test update of all fields within a single request

yuliiahonchar replied:
done

yuliiahonchar committed:
separate Session POJO and AuthenticationService
shorten ResponseUtils method names, move UpdateUserPositiveTest to update directory

yuliiahonchar committed:
Merge branch 'main' into update-user-tests

yuliiahonchar committed:
change parsing errorBody from gson to jackson
